rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,205 @@
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system.
+- Once a new EKS cluster is found, associate our OIDC provider with it.
+- Using the associated provider, generate an access token and install the kube agent through the Helm chart.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI.
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Identity Provider Association
+
+AWS EKS supports a method of authentication that involves associating an OIDC identity provider with the cluster.
+Such an associated identity provider can issue access tokens (JWT) that grant access to the EKS cluster groups and users.
+Since Teleport already supports OIDC providers, we can easily use this approach. It will require slight
+modifications to the AWS OIDC integration code to allow it to issue JWTs for an EKS cluster. To associate an OIDC provider with an
+EKS cluster, the selected AWS integration needs to issue an API call `eks:AssociateIdentityProviderConfig`
+with the following configuration (includes example parameters):
+
+```yaml
+---
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+
+metadata:
+  name: eks-cluster-name
+  region: us-east-1
+
+identityProviders:
+  - name: TeleportProvider
+    type: oidc
+    issuerUrl: https://teleport.example.com
+    clientId: kubernetes
+    usernameClaim: teleport-eks-oidc-user
+    groupsClaim: teleport-eks-oidc-groups
+```
+
+To access the target EKS cluster in API mode, we will generate an access token using the associated OIDC provider.
+That token will grant access to the `system:masters` group, so the Helm chart installation will be able to set up
+all the required infrastructure for the Teleport kube agent, including service accounts.
+
+### Plan of implementation
+
+Implementation will be done in three main steps:
+
+1. Introducing a UI flow for the enrollment of selected EKS clusters using AWS CloudShell.
+2. Adding API-only enrollment for selected EKS clusters with the use of an associated OIDC identity provider.
+3. Adding the capability to set up automatic discovery and enrollment of EKS clusters.
+
+Each of these steps is functionally complete and allows building upon it to implement the next step.
+
+## Product Usage
+
+We will add a new PostHog event to track the usage of the Discover flow for EKS.
```

Comment:
When we get to auto-discovery, we should also add an event that fires when an EKS fetcher triggers so we can see usage. @marcoandredinis is adding one currently for RDS discovery.